### ISSUE TITLE
ModuleAPI: Provide mechanism for modules to perform async initialisation (that is allowed to fail)

### ISSUE THREAD

**Issue by buffless-matt** [matrix-org/synapse#13409]
To carry over what was raised [here](https://matrix.to/#/!vcyiEtMVHIhWXcJAfl:sw1v.org/$LUehEb3JKlNza-6ymO-Qwjd5spM4v9OeFr1wdsTjd2A):

> Currently a module is instantiated via its `__init__` method, which (as a magic method) shouldn't really be async (as far as I know). What if we have a use-case that involves asynchronous initialisation of a module? E.g. Let's say we need to translate some non-local room aliases into room IDs on module initialisation and want to report an error (to Synapse) if that fails.

When I say "allowed to fail" (in the title), I mean similarly to how a module's `parse_config` static method may fail in a [controlled manner](https://github.com/matrix-org/synapse/blob/583f22780f44157c50bc2dc5c242e88cc18c7886/synapse/util/module_loader.py#L61-L66).

**Comment by erikjohnston** [comment]
There's a bit to unpack here:

First, during initialisation we don't have an async runtime, so we can't easily support this. We could defer initialisation of (async?) modules until after we've started the runtime, but at that point Synapse will have successfully started and its a bit late to bail out.

Secondly, if you just want to query the DB then what Synapse internals do is to query the DB synchronously, which would require the module API exposing the ability to get a database connection during initialisation.

Finally, the way that existing modules do this is spawn an async task in `__init__` that goes and calculates the necessary information. Any requests to the plugin that requires that info then waits until the aync task is finished. This works reasonably well, but doesn't allow you to abort Synapse startup if that async task fails (as it only starts running once Synpase has started)

**Comment by buffless-matt** [comment]
> First, during initialisation we don't have an async runtime, so we can't easily support this.

Bummer.

> Secondly, if you just want to query the DB then what Synapse internals do is to query the DB synchronously, which would require the module API exposing the ability to get a database connection during initialisation.

Seems plausible.

> Finally, the way that existing modules do this is spawn an async task in __init__ that goes and calculates the necessary information. Any requests to the plugin that requires that info then waits until the aync task is finished.

Could you please point me to a concrete example of this (so that I can see whether the pattern is suitable for my use-case)? It sounds like this is what I'm doing now anyway, but it would be useful to compare.

> This works reasonably well, but doesn't allow you to abort Synapse startup if that async task fails (as it only starts running once Synpase has started)

Yeah, that's the biggest motivator for creating this issue. A possible alternative would be some other means for a module user to easily determine whether the module "took off safely" or whether it "crashed on take-off". Thus far I've been resorting to manually inspecting the log file. Feasible, just not the same as a `ConfigError` being caught during a module's `parse_config` static method.

**Comment by erikjohnston** [comment]
> Could you please point me to a concrete example of this (so that I can see whether the pattern is suitable for my use-case)? It sounds like this is what I'm doing now anyway, but it would be useful to compare.

I'm failing to find any now. I think they might have been refactored to work differently.

> Yeah, that's the biggest motivator for creating this issue. A possible alternative would be some other means for a module user to easily determine whether the module "took off safely" or whether it "crashed on take-off". Thus far I've been resorting to manually inspecting the log file. Feasible, just not the same as a ConfigError being caught during a module's parse_config static method.

I guess its plausible for us to delay the health checks being marked as "OK" until the modules have finished loading.

--

I am a bit nervous about this whole idea in general tbh: async code during initialisation is a bit of a red flag, as you really *don't* want Synapse to fail to start due to some temporary blip. Synapse can get restarted randomly, and if it fails to start due to a random blip then there will be an outage. E.g. to take your example of resolving remote room aliases, what happens if that server is down or the aliases have been deleted?

**Comment by erikjohnston** [comment]
No worries! If it turns out there is a use case feel free to reopen :+1:

**Comment by erikjohnston** [comment]
Side thought: perhaps adding a prometheus metric that you can then alert on might give you want you want?

**Comment by richvdh** [comment]
> First, during initialisation we don't have an async runtime, so we can't easily support this. We could defer initialisation of (async?) modules until after we've started the runtime, but at that point Synapse will have successfully started and its a bit late to bail out.

I'm not entirely sure I agree with all this. The startup sequence is async, and afaict there's nothing stopping us doing async tasks before we tell systemd that we've started, or start the listener for `/health`. ([Here](https://github.com/matrix-org/synapse/blob/v1.63.0/synapse/app/_base.py#L464), for example.)

What *is* true is that, if you run Synapse in '--daemonize' mode, it will fork before it runs any async tasks... but it also forks before it opens the listening ports, so it's fairly badly broken anyway. 

Basically "Synapse has successfully started" is not terribly well defined.

Notwithstanding that, your other points about this being of questionable value are fair.

**Comment by babolivier** [comment]
This issue sounds like a duplicate of https://github.com/matrix-org/synapse/issues/12433

**Comment by buffless-matt** [comment]
> This issue sounds like a duplicate of https://github.com/matrix-org/synapse/issues/12433...

@babolivier this is definitely a duplicate of yours. Sorry I missed it.

**Comment by babolivier** [comment]
No worries :)
